### PR TITLE
Manually revert worker's goroutine 98a1ef45

### PR DESCRIPTION
**Pull request by miekg**
Manually revert the worker model.

### CODE REVIEW

**Comment by miekg** [comment]
ok,  much cleaner than #927, but failing a test; probably missed a `defer` somewhere that did a cleanup

**Comment by miekg** [comment]
ok, failing because of wait group stuff

File: server.go [hunk 1]
```diff
@@ -477,9 +408,7 @@ func (srv *Server) serveTCP(l net.Listener) error {
 		srv.NotifyStartedFunc()
 	}
 
-	var wg sync.WaitGroup
```
Comment (tmthrgd) [inline server.go]: This `sync.WaitGroup` is still needed for the graceful shutdown code.

File: server.go [hunk 2]
```diff
@@ -522,9 +446,7 @@ func (srv *Server) serveUDP(l *net.UDPConn) error {
 		reader = srv.DecorateReader(reader)
 	}
 
-	var wg sync.WaitGroup
```
Comment (tmthrgd) [inline server.go]: Ditto here.

File: server.go [hunk 3]
```diff
@@ -570,8 +489,6 @@ func (srv *Server) serve(w *response) {
 	if w.udp != nil {
 		// serve UDP
 		srv.serveDNS(w)
-
-		w.wg.Done()
```
Comment (tmthrgd) [inline server.go]: Ditto here.

File: server.go [hunk 4]
```diff
@@ -583,8 +500,6 @@ func (srv *Server) serve(w *response) {
 		srv.lock.Lock()
 		delete(srv.conns, w.tcp)
 		srv.lock.Unlock()
-
-		w.wg.Done()
```
Comment (tmthrgd) [inline server.go]: Ditto here too.

**Comment by tmthrgd** [comment]
I can cleanup some of these functions and get `msg` off of `response` in a follow-up PR. The server code was notably cleaner pre-worker.

**Comment by miekg** [comment]
all done. I'll to the msg untangling as well on top of this one.

**Comment by miekg** [comment]
#929 remove `msg` from response, this leads to a `nil` being passed to `serve()` which is somewhat ugly, but I can remember how that did look.

**Comment by tmthrgd** [comment]
@mielg Can this PR be retitled so it’s more obvious in the git history?

**Pull request merged**